lib/web/app/session.go

```diff
@@ -0,0 +1,173 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package app
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/tlsca"
+
+	"github.com/gravitational/trace"
+	"github.com/gravitational/ttlmap"
+
+	"github.com/gravitational/oxy/forward"
+	"github.com/sirupsen/logrus"
+)
+
+// session holds a request forwarder and web session for this request.
+type session struct {
+	// fwd can rewrite and forward requests to the target application.
+	fwd *forward.Forwarder
+	// ws represents the services.WebSession this requests belongs to.
+	ws services.WebSession
+}
+
+// newSession creates a new session.
+func (h *Handler) newSession(ctx context.Context, ws services.WebSession) (*session, error) {
+	// Extract the identity of the user.
+	certificate, err := tlsca.ParseCertificatePEM(ws.GetTLSCert())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	identity, err := tlsca.FromSubject(certificate.Subject, certificate.NotAfter)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Query the cluster this application is running in to find the public
+	// address and cluster name pair which will be encoded into the certificate.
+	clusterClient, err := h.c.ProxyClient.GetSite(identity.RouteToApp.ClusterName)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	accessPoint, err := clusterClient.CachingAccessPoint()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	application, server, err := Match(ctx, accessPoint, MatchPublicAddr(identity.RouteToApp.PublicAddr))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Create a rewriting transport that will be used to forward requests.
+	transport, err := newTransport(&transportConfig{
+		proxyClient:  h.c.ProxyClient,
+		accessPoint:  h.c.AccessPoint,
+		cipherSuites: h.c.CipherSuites,
+		identity:     identity,
+		server:       server,
+		app:          application,
+		ws:           ws,
+		clusterName:  h.clusterName,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	fwd, err := forward.New(
+		forward.RoundTripper(transport),
+		forward.Logger(h.log))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &session{
+		fwd: fwd,
+		ws:  ws,
+	}, nil
+}
+
+// sessionCache holds a cache of sessions that are used to forward requests.
+type sessionCache struct {
```

Comment:
I wonder if you can somehow reuse sessionCache from lib/srv/app package since it appears to be the same.